cloud-controller-manager/do/firewall_controller.go

```diff
@@ -156,20 +171,52 @@ func NewFirewallController(
 	return fc
 }
 
+// obj could be an *v1.Service, or a DeletionFinalStateUnknown marker item.
+func (fc *FirewallController) enqueueService(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+	fc.queue.Add(key)
+}
+
 // Run starts the firewall controller loop.
-func (fc *FirewallController) Run(ctx context.Context, stopCh <-chan struct{}, fwReconcileFrequency time.Duration) {
-	wait.Until(func() {
-		err := fc.observeRunLoopDuration(ctx, stopCh, fwReconcileFrequency)
-		if err != nil {
-			klog.Errorf("failed to run firewall controller loop: %v", err)
-		}
-	}, fwReconcileFrequency, stopCh)
+func (fc *FirewallController) Run(threadiness int, stopCh <-chan struct{}, fwReconcileFrequency time.Duration) {
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(fc.runWorker, fwReconcileFrequency, stopCh)
+	}
+
```

Comment:
We should [make sure that the queue is shut down](https://github.com/kubernetes/client-go/blob/ba26c4506d39dece49d00e969d03794e8089b8dc/examples/workqueue/main.go#L119-L120).